progenitor-impl/src/lib.rs

```diff
@@ -262,6 +266,43 @@ impl Generator {
 
         let types = self.type_space.to_stream();
 
+        let extra_impl = TokenStream::from_iter(
+            self.forms
+                .iter()
+                .map(|type_id| {
+                    let typ = self.get_type_space().get_type(type_id).unwrap();
+                    let form_name = typ.name();
+                    let td = typ.details();
+                    let TypeDetails::Struct(tstru) = td else { unreachable!() };
+                    let properties = indexmap::IndexMap::<&'_ str, _>::from_iter(
+                        tstru
+                            .properties()
+                            .filter_map(|(prop_name, prop_id)| {
+                                self.get_type_space()
+                                    .get_type(&prop_id).ok()
+                                    .map(|prop_typ| (prop_name, prop_typ))
+                            })
+                        );
+                    let properties = syn::punctuated::Punctuated::<_, syn::Token![,]>::from_iter(
+                        properties
+                            .into_iter()
+                            .map(|(prop_name, prop_ty)| {
+                                let ident = quote::format_ident!("{}", prop_name);
+                                quote!{ (#prop_name, &self. #ident) }
+                            }));
+
+                    let form_name = quote::format_ident!("{}",typ.name());
+
+                    quote! {
+                        impl #form_name {
+                            pub fn as_form<'f>(&'f self) -> impl std::iter::Iterator<Item=(&'static str, &'f [u8])> {
+                                [#properties]
+                                    .into_iter()
+                                    .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_slice())))
```

Comment:
Does this mean that all `val` have to be strings?

Reply:
No, not at all. It means all values have to be convertible to a stream of bytes. The representation is limiting and should be improved, i.e. it only allows for basic usage `key` but not for keys derived from objects where the keys end up as `foo[bar].sub`

Reply:
I see. Are there types we might encounter here other than `String` and `Vec[u8]` that satisfy this?

Reply:
Any type really, such that we could call `as_form` for nested types. That was the direction I wanted to go into. Hence avoiding the direct expansion. There are a few pitfalls, i.e. metadata desc such as mime type and filename that are commonly expected but not defined in the openapi spec. I am not yet sure how to handle this.

Reply:
Note that the PR here is in a state of rough draft/idea/make my immediate needs meet, so I am open to changing direction as you see fit.

Reply:
Gentle ping on this one.